### PR TITLE
Added adaptive layout modes

### LINKED ISSUE #2893: Automatically switch to grid mode when the majority of items are photos
**Is your feature request related to a problem? Please describe.**
When I open a folder that contains mostly pictures, I wish I didn't have to manually switch to grid view.

**Describe the solution you'd like**
It would be helpful if when the majority of items in a folder (like with an exception of 5 or something) were pictures, the layout could automatically switch to the small grid view and back again when you exit.

**Describe alternatives you've considered**
I would change it myself each time I enter/exit.

**Additional context**
Windows 10 Version 20H2 (OS Build 19042.685)

Files:
App Version: 0.23.1.0

### PR DESCRIPTION
**Pull request by d2dyno1**
This PR adds Adaptive Layout Mode. I'd like to hear your feedback on current logic of changing Layout Modes.

**Things yet to do**
- [x] Move Adaptive Layout option to settings
- [x] Make Adaptive Layout select the layout modes
- [x] Fix disabling feature for folder layout save
- [x] Add desktop.ini reading
- [x] Remember preferred gridview sizemode
- [x] https://github.com/files-community/Files/pull/3526#issuecomment-780066405
- [x] Replace "Re-enable Adaptive Layout" button with something different..

**Limitations**
- ~~Feature "Enable individual preferences for individual directories" is disabled when Adaptive Layout is enabled~~

<br/>

~~To enable Adaptive Layout Mode, select it from Display Options~~

image
~~-> It's been moved to settings~~
It is enabled by default and cannot be disabled

Closes #2893

### CODE REVIEW

**Comment by d2dyno1** [comment]
Me not being able to get the maths right for 40 minutes straight:

image

**Comment by yaichenbaum** [comment]
@d2dyno1 Instead of having a toggle on the flyout, can you make this this a setting?

**Comment by d2dyno1** [comment]
> @d2dyno1 Instead of having a toggle on the flyout, can you make this this a setting?

But what will other buttons do if they will always be overriden?
image

**Comment by yaichenbaum** [comment]
@d2dyno1 The preferred way would be to use the same behavior that's found in File Explorer.

**Comment by d2dyno1** [comment]
It looks like this logic behaves similarly to File Explorer, or am I missing something?

@yair100

File: Files/ViewModels/ItemViewModel.cs
Comment (duke7553) [inline Files/ViewModels/ItemViewModel.cs:1011]: Are we sure that `item.FileExtension` will always return a lowercase string? If not, we should compare with `.Equals()` instead.

**Comment by d2dyno1** [comment]
@yair100 ready for review! :shipit:

**Comment by yaichenbaum** [comment]
@d2dyno1 How is adaptive layout meant to work?

File: Files/Helpers/AdaptiveLayoutHelpers.cs
Comment (yaichenbaum) [inline Files/Helpers/AdaptiveLayoutHelpers.cs:168]: Instead of comparing the extension, what about comparing the file type?

File: Files/Views/SettingsPages/FilesAndFolders.xaml
Comment (yaichenbaum) [inline Files/Views/SettingsPages/FilesAndFolders.xaml:77]: The text here is a bit long, do you have any ideas how it can be simplfied?

**Comment by d2dyno1** [comment]
> @d2dyno1 How is adaptive layout meant to work?

You can find the logic here: https://github.com/files-community/Files/pull/3526/files#diff-1fc2ffb3299cf773313a1100cee887362f6a9b04450aa620fdc1af7656ddffbeR67

Basically, it looks for desktop.ini to get the information about the folder, if no data in desktop.ini is present, it makes some predictions based on the items in the directory - e.g.:
Directory contains mostly images -> TilesView
Directory contains images and/or videos -> GridView

File: Files/Views/SettingsPages/FilesAndFolders.xaml
Comment (yaichenbaum) [inline Files/Views/SettingsPages/FilesAndFolders.xaml:76]: Can you add strings to the en-us resource file once the text is figured out?

**Review by yair100 — CHANGES_REQUESTED** [review]

**Comment by yaichenbaum** [comment]
> > @d2dyno1 How is adaptive layout meant to work?
> 
> You can find the logic here: https://github.com/files-community/Files/pull/3526/files#diff-1fc2ffb3299cf773313a1100cee887362f6a9b04450aa620fdc1af7656ddffbeR67
> 
> Basically, it looks for desktop.ini to get the information about the folder from, if no data in desktop.ini is present, it makes some predictions based on the items in the directory - e.g.:
> Directory contains mostly images -> TilesView
> Directory contains images and/or videos -> GridView

I assume that once a user overrides that for a directory, adaptive mode will ignore that directory. Is that correct?

File: Files/Views/SettingsPages/FilesAndFolders.xaml
Comment (d2dyno1) [reply Files/Views/SettingsPages/FilesAndFolders.xaml:77]: I'll think of that. Honestly, #2905 would be perfect fit here

**Comment by d2dyno1** [comment]
> I assume that once a user overrides that for a directory, adaptive mode will ignore that directory. Is that correct?

It does not do that. Currently when a gridview is selected by the Adaptive Layout, user can only override the grid size.

Sounds promising though, I'll see what could be done

@yair100

**Comment by yaichenbaum** [comment]
> It does not do that. Currently when a gridview is selected by the Adaptive Layout, user can only override the grid size.

Ok, before merging, we should change this behavior to let the user override it completely. The behavior in File Explorer is good so we should try to match it.

**Comment by d2dyno1** [comment]
I've edited some logic in `a025af0`

@yair100

**Comment by yaichenbaum** [comment]
@d2dyno1 Being that the user can override a directory, we can probably remove the setting to turn off this feature.

**Comment by d2dyno1** [comment]
> @d2dyno1 Being that the user can override a directory, we can probably remove the setting to turn off this feature.

There should be still an option to disable this. Currently, when user overrides the layout, it is only overriden in that specific directory and other folders are not affected by it unless they're also overriden :shipit: 

@yair100

**Comment by yaichenbaum** [comment]
@d2dyno1 The behavior in File Explorer is pretty good in this area, I am not sure we need extra customization. Adding @duke7553 for a second opinion.

**Comment by d2dyno1** [comment]
We should keep this enabled by default. There are some people including me which would like to have this feature disabled for some scenarios. If you ask me, we should keep the switch for it as it is now

@yair100

**Comment by gave92** [comment]
Help! Is this just me? xD

https://user-images.githubusercontent.com/9673091/108111834-8cd2df00-7095-11eb-885b-1622e8a2d785.mp4

**Comment by d2dyno1** [comment]
> Help! Is this just me? xD

oof, looks like we have a new issue. I wonder what might have caused that 🤔

**Comment by d2dyno1** [comment]
@gave92 fixed!

**Comment by gave92** [comment]
Previous issue fixed, new issue is:
```cs
```

**Comment by d2dyno1** [comment]
The logic of remembering where the adaptive layout is overridden seems broken, will have to redo it :/

**Comment by d2dyno1** [comment]
@gave92 Am I going crazy or is `UpdateLayoutPreferencesForPath()` not saving the bool `AdaptiveLayoutDisabledOverride`?!

**Comment by d2dyno1** [comment]
> Previous issue fixed, next issue is:

@gave92 fixed in `035af65`

**Comment by d2dyno1** [comment]
> [Side note: do we really have to read stuff from the ini file? I have yet to find a desktop.ini file on my system that actually contains the view mode..]

@gave92 desktop.ini will be automatically created if not present, when edited the folder optimization settings, as per [this](https://github.com/files-community/Files/issues/2893#issuecomment-757116130) comment. Just make sure you have "Hide protected system files" feature turned off.

We are reading from the .ini file to check if the user has saved directory preferences.

**Comment by gave92** [comment]
Aaah it only gets created when you set the folder type, *not* when changing the folder layout mode. Pity.

**Comment by yaichenbaum** [comment]
@d2dyno1 After talking it over with @duke7553, we decided that there shouldn't be a setting here as it's not something most users will care about.

**Comment by d2dyno1** [comment]
@yair100 ok. I hid it so it's still here if you ever change your mind. Now it is enabled by default and cannot be turned off

**Comment by yaichenbaum** [comment]
@d2dyno1 Can you remove the xaml and strings for the UI completly? I would still keep the setting in SettingsViewModel, but the xaml is a little extra.

File: Files/UserControls/StatusBarControl.xaml
Comment (yaichenbaum) [inline Files/UserControls/StatusBarControl.xaml:162]: Do you think this button is redundant most of the time? Do you have any other ideas what we can do here?

File: Files/ViewModels/SettingsViewModel.cs
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModel.cs:435]: What does this setting do?

Comment (d2dyno1) [reply Files/ViewModels/SettingsViewModel.cs:435]: This setting is managed by code and user can't override it by themselves. It's purpose is to save which gridview size is preferred. I put it in settings because I had no other idea where to save it.
As always, I'm opened to suggestions

Comment (yaichenbaum) [reply Files/ViewModels/SettingsViewModel.cs:435]: The setting itself is fine, however, it should be together with the other layout settings, not in the FilesAndFolders region. Can you also add a description for documentation?

Comment (d2dyno1) [reply Files/ViewModels/SettingsViewModel.cs]: @yair100 I've removed this setting in https://github.com/files-community/Files/pull/3526/commits/060e41e1b5dc43d3e4058d46e9afa3e7643032b5 - different size modes are now remembered for every directory!

**Comment by d2dyno1** [comment]
@yair100 I'm not sure how to implement that shift click to re-enable Adaptive Layout.

Could you help me with that?

**Comment by gave92** [comment]
I haven't had the time to review this so I might be wrong:
If the problem is to reenable adaptive layout after the user overrode it for a folder I think a good solution would be to use the layout flyout buttons as below:

1) User clicks on details view button -> details view overrides adaptive mode and the button gets selected
image

2) User click again on detail view button ->  adaptive mode is restored and the button gets deselected
image

So the button would show the current layout mode if adaptive layout is overridden or no selection when adaptive layout is enabled for the folder

**Comment by d2dyno1** [comment]
@yair100 @gave92 

In https://github.com/files-community/Files/pull/3526/commits/bc29baeceb7cdf7a688ff8611b8a4609e9407f01 I've made it so when user presses the layout mode option (that's currently selected by user, not by Adaptive Layout) for the second time, it re-enables Adaptive Layout.
Let me know if you like this solution and in next commit I'll remove the button.


https://user-images.githubusercontent.com/53011783/109692901-557e2b00-7b89-11eb-9946-5eb31ea384ca.mp4

**Comment by d2dyno1** [comment]
Ready for review!

**Comment by yaichenbaum** [comment]
@d2dyno1 Can you solve the merge conflicts?

**Comment by d2dyno1** [comment]
@yair100 yessir 👍

File: Files.Launcher/Program.cs
Comment (yaichenbaum) [inline Files.Launcher/Program.cs:604]: Was this removed on purpose?

File: Files/ViewModels/SettingsViewModel.cs
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModel.cs:405]: Can this be removed?

File: Files.Launcher/Program.cs
Comment (d2dyno1) [reply Files.Launcher/Program.cs]: @yair100 That's an issue while resolving merge conflicts. I have fixed it and restored the code

**Review by yair100 — DISMISSED** [review]

**Comment by d2dyno1** [comment]
Looking for feedback on the Adaptive Layout layout switching decision logic.

@yair100 I'd like for the build that ships with Adaptive Layout to stay for a bit longer in insider channel than usual so we can collect necessary feedback from insiders on this feature to meet standards of productivity for users before it is pushed to the main branch.

**Comment by d2dyno1** [comment]
This feature must be perfected since it can't be disabled.

**Comment by yaichenbaum** [comment]
@d2dyno1 We can have a longer testing period to help iron out issues.

**Comment by gave92** [comment]
> @gave92 Am I going crazy or is UpdateLayoutPreferencesForPath() not saving the properties - specifically AdaptiveLayoutDisabledOverride?!

You were right @d2dyno1, layout preferences were not getting saved on occasion (after #3716), sorry about that.

**Comment by d2dyno1** [comment]
> You were right @d2dyno1, layout preferences were not getting saved on occasion (after #3716), sorry about that.

phew!! I thought I was going insane 😅

**Review by gave92 — APPROVED** [review]

**Review by yair100 — APPROVED** [review]
Great work!

**Comment by gave92** [comment]
@candrapersada If Files crashes please open a new issue and upload the app's debug logs.

**Pull request merged**
